File: daily_python/3217_Delete_Nodes_From_Linked_List_Present_in_Array.py
import os
import sys
from typing import List, Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def create_linked_list(nodes):
    if not nodes:
        return None
    root = ListNode(nodes[0])
    temp = root
    for val in nodes[1:]:
        temp.next = ListNode(val)
        temp = temp.next
    return root

def print_list(root: Optional[ListNode]):
    while root:
        print(root.val, end = ' ')
        root = root.next
    print()

class Solution:
    def modifiedList(self, nums: List[int], head: Optional[ListNode]) -> Optional[ListNode]:
        remove = set(nums)
        pseudo = ListNode(0, head)
        temp = pseudo
        ptr = head
        while ptr:
            if ptr.val not in remove:
                temp.next = ptr
                temp = temp.next
            ptr = ptr.next
        temp.next = None
        return pseudo.next
    
def readDataSet(filename):
    dataset = []
    with open(filename) as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            nodes = list(map(int, lines[1].split('=')[1].strip()[1:-2].split(',')))
            dataset.append((nums, nodes))
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums, nodes in dataset:
        head = create_linked_list(nodes)
        results.append(solution.modifiedList(nums, head))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : ', end = '')
        print_list(result)

File: daily/2024/3217_Delete_Nodes_From_Linked_List_Present_in_Array.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* create_list(vector<int>& nodes){
    if (nodes.empty()) return nullptr;
    ListNode* root = new ListNode(nodes[0]);
    ListNode* temp = root;
    for (int i = 1; i < nodes.size(); i++){
        temp->next = new ListNode(nodes[i]);
        temp = temp->next;
    }
    return root;
}

void print_list(ListNode* root){
    while(root){
        cout << root->val << " ";
        root = root->next;
    }
    cout << endl;
    return;
}

class Solution {
public:
    ListNode* modifiedList(vector<int>& nums, ListNode* head) {
        unordered_set<int> s;
        for (int i : nums) s.insert(i);
        ListNode* pseudo = new ListNode(0, head);
        ListNode* temp = pseudo;
        ListNode* ptr = head;
        while(ptr != nullptr){
            if (s.find(ptr->val) == s.end()){
                temp->next = ptr;
                temp = temp->next;
            }
            ptr = ptr->next;
        }
        temp->next = nullptr;
        return pseudo->next;
    }
};

int main(){
    vector<int> nums = {1,2,3};
    vector<int> nodes = {1,2,3,4,5};
    // vector<int> nums = {1};
    // vector<int> nodes = {1,2,1,2,1,2};
    // vector<int> nums = {5};
    // vector<int> nodes = {1,2,3,4};
    ListNode* head = create_list(nodes);
    Solution S;
    ListNode* ans = S.modifiedList(nums, head);
    print_list(ans);
    return 0;
}